src/autosave.js

```diff
@@ -0,0 +1,288 @@
+/**
+ * Copyright (c) 2016 - 2017, CKSource - Frederico Knabben. All rights reserved.
+ */
+
+/**
+ * @module autosave/autosave
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+import PendingActions from '@ckeditor/ckeditor5-core/src/pendingactions';
+import DomEmitterMixin from '@ckeditor/ckeditor5-utils/src/dom/emittermixin';
+
+/* globals window */
+
+/**
+ * Autosave plugin provides an easy-to-use API to save the editor's content.
+ * It watches {@link module:engine/model/document~Document#event:change:data change:data}
+ * and `Window:beforeunload` events and calls the {@link module:autosave/autosave~SaveProvider#save} method.
+ *
+ * 		ClassicEditor
+ *			.create( document.querySelector( '#editor' ), {
+ *				plugins: [ ArticlePluginSet, Autosave ],
+ *				toolbar: [ 'heading', '|', 'bold', 'italic', 'link', 'bulletedList', 'numberedList', 'blockQuote', 'undo', 'redo' ],
+ *				image: {
+ *					toolbar: [ 'imageStyle:full', 'imageStyle:side', '|', 'imageTextAlternative' ],
+ *				}
+ *			} )
+ *			.then( editor => {
+ *				const autosave = editor.plugins.get( Autosave );
+ *				autosave.provider = {
+ *					save() {
+ *						const data = editor.getData();
+ *
+ *						// Note: saveEditorsContentToDatabase function might be async and return a promise to the saving action.
+ *						return saveEditorsContentToDatabase( data );
+ *					}
+ *				};
+ *			} );
+ */
+export default class Autosave extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'Autosave';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ PendingActions ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * Provider is an object with the `save()` method. That method will be called whenever
```

Comment:
It's an artificial, over-corporate naming for me. What does this provider provide? A save callback. So it's a save callback provider. In other words... it's a save callback.

Besides, we already use the word adapter for the same thing – defining a way to communicate with an external service. So, if we need any funky noun here, "adapter" would do better.

---

If my point isn't clear, I recommend the next 10 minutes of https://youtu.be/ZsHMHukIlJY?t=23m41s

And this:

![image](https://user-images.githubusercontent.com/156149/41409660-34fb784e-6fd7-11e8-8f8d-77dee40b9bc0.png)

Reply:
> It's an artificial, over-corporate naming for me. What does this provider provide? A save callback. So it's a save callback provider. In other words... it's a save callback.

IMO object that provides a callback isn't a callback and calling it callback would be misleading. I didn't want to use a fancy name. For me, an Adapter would be something that has to adapt to some other interface, so someone can create an adapter that would fit our interface. But from the `Autosave` plugin's point of view, it doesn't need any adapter, it needs some object that provides a `save` method, thus I found the `SaveProvider` name much better.

Either way, I can change it to follow our code naming conventions.

Reply:
> it doesn't need any adapter, it needs some object that provides a save method, thus I found the SaveProvider name much better.

It does not need some object. It needs the save method. I wasn't clear – I don't see (now) the need for that object.

Besides, I've got a new question – why can't the autosave plugin be configurable through the editor's config?

Reply:
> It does not need some object. It needs the save method. I wasn't clear – I don't see (now) the need for that object.

So maybe it would be better to have only the `Autosave#save` property, which would be a callback for the `Autosave`. For now, there's an object, which is unnecessary indeed. cc @pjasiun

> I recommend the next 10 minutes of youtu.be/ZsHMHukIlJY?t=23m41s

It's a really good talk :) Thanks for sharing.

> Besides, I've got a new question – why can't the autosave plugin be configurable through the editor's config?

We can add a configuration for it through the `{ autosave: { save } }` option. It's a good idea.

Reply:
> Besides, I've got a new question – why can't the autosave plugin be configurable through the editor's config?

Because one may want to create a plugin which will provide a specific integration, for instance, `S3Autosave` or `DropboxAutosave` it will be more complicated to do with a config. It is a similar feature to upload repository which needs an adapter to works.

When it comes to names I am fine with both adapter and provider.

The reason why I am not sure about just a `save()` callback, is that it does not show in any way that this it the special method you need to overwrite to make the plugin works. It is strange from both sides: for future `autosave` developers: that there a method which is used but is not defined, need to be injected. For people who integrate autosave, it looks like overwriting a random method. When this method is wrapped in `adapter` or `provider` it becomes of obvious that it is some special method to be overwritten.

Reply:
> Because one may want to create a plugin which will provide a specific integration

That's a special case. The basic usage (end-developer oriented) will be through the editor config. But I agree that we can support config + plugin's property.

> When this method is wrapped in adapter or provider it becomes of obvious that it is some special method to be overwritten.

OK, but let's just be consistent with the upload adapter and call this property an `adapter`.

Reply:
Can I assume, that the config won't change after initialization, so I can store a config for the autosave? Or should I be aware of `editor.config.set`? Do we even support config change anywhere?